### PR TITLE
refactor message parsing

### PR DESCRIPTION
**Pull request by hwchen**
Hi! I saw your call for feedback on twitter, if you don't mind I have some suggestions here.

- refactored 'message' module
  - changed `from_string` to return `Option` instead of `Result` for now, since there's nothing internally returning `Result`.  If you want to pass along more info on error, you'd probably have to start defining an `Error` enum.
  - More clearly align lifetime of `Message` with the line it's referencing.
  - use `chars`, which might be more idiomatic? This is probably not a big deal.
  - remove unwrap. The first is changed to `and_then`, since you don't want to early return. The second is turned into `?` for now, since it looks like it could early return.

- Changed `handle_lines` to return `Box<dyn std::error::Error>`, since there's two types of errors that can be returned internally.
  - since `Message::from_string` returns an `Option` for now, I use `ok_or` to turn that into a `Result`

Three other overall notes:
- error handling is probably the next big step. Let me know if you have any questions about it, there's currently a lot to navigate.
- You wrote a `from_string` method; there's a `FromStr` trait in `std`, which I would generally recommend implementing.
- You use `&str` for `nick` and `label` in your `LineHandler` and `Session` structs; I think that this may complicate lifetimes down the road for not much gain. I don't think it hurts for these structs to own their `label` and `nick` data and I don't think there's much any performance gain in referencing. Using `&str` does make sense for `Message` though.

Let me know if you have any other questions, thanks for writing this tutorial!

### CODE REVIEW

File: src/irc/message.rs [hunk 1]
```diff
@@ -4,27 +4,35 @@ pub struct Message<'a> {
...
     let mut parts = line.split(' ');
-    let prefix = if line.get(0..1).unwrap() == ":" {
-      parts.next()
-    } else {
-      None
-    };
 
-    Ok(Message {
+    let prefix = line.chars().nth(0)
+        .and_then(|c| {
```
Comment (ultrasaurus) [inline src/irc/message.rs]: I like this!  I didn't know about [and_then](https://doc.rust-lang.org/rust-by-example/error/option_unwrap/and_then.html)

File: src/irc/message.rs [hunk 2]
```diff
@@ -4,27 +4,35 @@ pub struct Message<'a> {
...
+            if c == ':' {
+                parts.next()
+            } else {
+                None
+            }
+        });
+
+    let command = parts.next()?;
```
Comment (ultrasaurus) [inline src/irc/message.rs]: I thought to use `?` the function has to return a Result..?

Comment (hwchen) [reply src/irc/message.rs]: As of now on stable, `?` can be used with `Option`. https://blog.rust-lang.org/2017/11/22/Rust-1.22.html.

For returning an `Option` into a `Result`, there's this on [nightly](https://doc.rust-lang.org/std/option/struct.NoneError.html), or you can use [ok_or](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or)

File: src/irc/message.rs (hunk 1 above)
Comment (hwchen) [reply src/irc/message.rs]: Yeah, sometimes the combinators work out nicely (but sometimes they are a bit awkward, it depends on the specific use case)

**Comment by hwchen** [comment]
Yup, I'm referring to using `String` instead of `&str` in those structs.

**Pull request merged**